File: ios/Sources/IsekaiTerminalCoreLogic/TerminalScrollback.swift
import Foundation

/// Phase 1F-4(#51): ライブの`update`とスクロールバックの行から表示用の`ScreenUpdate`を
/// 合成する。Android版`tools.isekai.terminal.ui.synthesizeDisplayUpdate`
/// (`TerminalScrollback.kt`、`TerminalScreen.kt`の`displayUpdate`から呼ばれる)と対称
/// (タスク#46でAndroid側もこの合成ロジックを独立関数へ抽出した)。`scrollOffset == 0`かつ
/// `showingScrollback`が`false`のときだけライブをそのまま返す——`scrollOffset == 0`は
/// 「ライブ画面表示」と「scrollback最新行(row=0)表示」の両方を指しうるため、
/// `showingScrollback`で明示的に区別する(タスク#79: 検索結果のrow=0へジャンプする際、
/// `scrollOffset`を変えずにscrollback最新行を合成表示させるために追加。呼び出し元
/// `TerminalView.swift`の`showingScrollback`参照、Android版`TerminalScreen.kt`と対称)。
/// `scrollbackCells`の件数が`cols * rows`と一致しない場合(未取得・セッション未確立等)も
/// ライブへフォールバックする。
///
/// スクロールバック表示中はカーソルを画面外(`cursorRow = update.rows`)に隠す
/// (Android版と同じ、ライブでない行にカーソルを描くのは意味がないため)。
public func synthesizeDisplayUpdate(
    live update: ScreenUpdate,
    scrollOffset: UInt32,
    scrollbackCells: [CellData],
    showingScrollback: Bool = false
) -> ScreenUpdate {
    guard scrollOffset > 0 || showingScrollback else { return update }
    let cols = Int(update.cols)
    let rows = Int(update.rows)
    guard cols > 0, rows > 0, scrollbackCells.count == rows * cols else { return update }
    return ScreenUpdate(
        // updateSeqはライブの値をそのまま引き継ぐ(この合成はdraw経路の
        // `computeDisplayUpdate`内でのみ使われ、`apply`のupdateSeqギャップ検出は通らない
        // ——値は実質inertだが、他フィールド同様ライブを忠実にミラーしておく)。
        updateSeq: update.updateSeq,
        cols: update.cols, rows: update.rows, cells: scrollbackCells,
        cursorRow: update.rows, cursorCol: 0,
        title: update.title,
        applicationCursorMode: update.applicationCursorMode,
        applicationKeypadMode: update.applicationKeypadMode,
        bracketedPasteMode: update.bracketedPasteMode,
        mouseReportingMode: update.mouseReportingMode,
        sgrMouseMode: update.sgrMouseMode,
        alternateScroll: update.alternateScroll,
        urxvtMouseMode: update.urxvtMouseMode,
        cursorVisible: update.cursorVisible,
        bellGeneration: update.bellGeneration,
        cursorShape: update.cursorShape,
        cursorBlink: update.cursorBlink,
        linkTable: update.linkTable,
        // Sixel(タスク#42)/Kitty graphics(タスク#53): scrollback表示中はライブ画面の画像
        // 配置を引き継がない(scrollbackセル自体は画像を保持しないテキストのみの
        // スナップショットのため、Android版`synthesizeDisplayUpdate`と同じ判断)。
        images: [],
        kittyKeyboardFlags: update.kittyKeyboardFlags,
        // スクロールバック合成は表示グリッド全体を差し替えるため全画面dirty(=nil)扱い。
        // ライブの`update.dirtyRows`はライブグリッドの行番号を指しており、scrollback表示
        // 行とは対応しないので引き継がない(タスク#102)。
        dirtyRows: nil
    )
}

/// タスク#66/#67: 検索バーの現在マッチ(`match`)のうち、実際に`scrollOffset`の位置へ
/// ハイライトとして描画してよいものだけを返すピュア関数。Android版
/// `tools.isekai.terminal.searchHighlightMatch`(`TerminalScreen.kt`)と対称
/// (タスク#79でAndroid側に続きiOS側もこの判断ロジックを独立関数へ抽出し、UIKit非依存の
/// `IsekaiTerminalCoreLogic`層でユニットテスト可能にした——以前は`TerminalView.swift`の
/// `body`内にインライン記述されていたため、`TerminalScreenViewTests`のクラッシュ確認
/// スモークテストでしか間接的にしか検証できなかった)。
///
/// `ScrollbackSearchMatch.row`は`scrollbackCells`と同じ規約("offset"がそのまま`row`)なので、
/// `scrollOffset`がその値と一致している間だけ実際に画面へ表示される。`scrollOffset == 0`は
/// 「ライブ画面表示」と「scrollback最新行(row=0)表示」の両方を指しうる(既存規約
/// `synthesizeDisplayUpdate`と`showingScrollback`参照)ため、`row == 0`のマッチは
/// `showingScrollback`が真の間(=実際にscrollback最新行を表示中)だけハイライトを許可する
/// (タスク#79: それ以外[ライブ画面表示中]にrow=0のマッチを誤ってハイライトしないための
/// ガード)。
public func searchHighlightMatch(
    _ match: ScrollbackSearchMatch?,
    scrollOffset: UInt32,
    showingScrollback: Bool
) -> ScrollbackSearchMatch? {
    guard let match, scrollOffset == match.row, match.row != 0 || showingScrollback else { return nil }
    return match
}

File: android/src/main/kotlin/tools/isekai/terminal/ui/TerminalScrollback.kt
package tools.isekai.terminal.ui

import uniffi.isekai_terminal_core.CellData
import uniffi.isekai_terminal_core.ScreenUpdate

/**
 * タスク#46: ライブの[live]とスクロールバックの行([scrollbackCells])から表示用の
 * [ScreenUpdate]を合成する。iOS版`TerminalScrollback.swift`の`synthesizeDisplayUpdate`と
 * 対称(このファイルはその移植元)。`scrollOffset <= 0`かつ[showingScrollback]が偽の
 * ときだけライブをそのまま返す——`scrollOffset == 0`は「ライブ画面表示」と
 * 「scrollback最新行(row=0)表示」の両方を指しうるため、[showingScrollback]で明示的に
 * 区別する(タスク#79: 検索結果のrow=0へジャンプする際、`scrollOffset`を変えずに
 * scrollback最新行を合成表示させるために追加。呼び出し元`TerminalScreen.kt`の
 * `showingScrollback`参照)。[scrollbackCells]の件数が`live.cols * live.rows`と一致しない
 * 場合(未取得・セッション未確立・リサイズ中の過渡状態等)もライブへフォールバックする。
 *
 * 検証対象の寸法は必ず[live]自身の`cols`/`rows`から導出すること(呼び出し側でCompose層が
 * 独自に計算したビューポート由来のcols/rowsを渡してはいけない)。`actions.onScrollbackCells`
 * へのリクエストも同じ[live]の寸法で行うべき(呼び出し元`TerminalScreen.kt`参照)——
 * Rust側`SessionCore.resize()`はチャネル送出成功時点で`screen_cols`を即時更新する一方、
 * `live`(直近の`ScreenUpdate`ブロードキャスト)側は非同期に遅れて追随するため、
 * Compose層が独自計算したcols/rowsとRust側の実際のスクロールバック行幅が過渡的に
 * 食い違いうる(Codexレビュー: タスク#46)。両方を[live]基準に揃えることで、
 * 「検証は通ったが返す`ScreenUpdate.cols/rows`と実際の`cells`サイズが食い違う」
 * (`SshTerminalCanvas`側のインデックス計算がずれ、最悪`IndexOutOfBoundsException`になる)
 * という事故を構造的に防ぐ。
 *
 * `ScreenUpdate`にフィールドを追加すると、この関数を含む位置引数コンストラクタ呼び出しは
 * 必ずコンパイルエラーになる(UniFFI生成型はデフォルト引数を持たない)ため「更新忘れ」自体は
 * コンパイラが防ぐ。本当に注意が必要なのは**意味論**: 新フィールドをここでライブから
 * そのまま引き継ぐべきか、blank値にすべきかを都度判断すること。例えば`bellGeneration`は
 * BEL通知用のRust側SSOTカウンタ(タスク#24)なので、スクロールバック表示中も直近のライブ値を
 * 落としてはいけない(=引き継ぐ)。一方`images`(Sixel/Kittyタスク#42・#53)はscrollbackセル
 * 自体が画像を保持しないテキストのみのスナップショットのため、意図的に空にする。
 */
fun synthesizeDisplayUpdate(
    live: ScreenUpdate,
    scrollOffset: Int,
    scrollbackCells: List<CellData>?,
    showingScrollback: Boolean = false,
): ScreenUpdate {
    if (scrollOffset < 0 || (scrollOffset == 0 && !showingScrollback)) return live
    val cols = live.cols.toInt()
    val rows = live.rows.toInt()
    // iOS版と同じくcols/rowsが0(未初期化画面等)の縮退ケースも明示的にライブへ
    // フォールバックする(`rows * cols == 0`のとき空のscrollbackCellsが偶然一致してしまう
    // のを避ける)。
    if (cols <= 0 || rows <= 0) return live
    if (scrollbackCells == null || scrollbackCells.size != rows * cols) return live
    return ScreenUpdate(
        // scrollback合成は必ず全画面dirty(下記 dirtyRows = null)なので updateSeq のギャップ判定に
        // 影響しないが、下敷きのライブフレームの連番をそのまま引き継いでおく。
        updateSeq = live.updateSeq,
        cols = live.cols,
        rows = live.rows,
        cells = scrollbackCells,
        cursorRow = live.rows, // カーソルは画面外に隠す(ライブでない行に描く意味が無いため)
        cursorCol = 0u,
        title = live.title,
        applicationCursorMode = live.applicationCursorMode,
        applicationKeypadMode = live.applicationKeypadMode,
        bracketedPasteMode = live.bracketedPasteMode,
        mouseReportingMode = live.mouseReportingMode,
        sgrMouseMode = live.sgrMouseMode,
        alternateScroll = live.alternateScroll,
        urxvtMouseMode = live.urxvtMouseMode,
        cursorVisible = live.cursorVisible,
        bellGeneration = live.bellGeneration,
        cursorShape = live.cursorShape,
        cursorBlink = live.cursorBlink,
        linkTable = live.linkTable,
        // Sixel(タスク#42)/Kitty graphics(タスク#53): scrollback表示中はライブ画面の
        // 画像配置を引き継がない(scrollbackセル自体は画像を保持しないテキストのみの
        // スナップショットのため、cursorVisible相当の考え方で画像も非表示にする)。
        images = emptyList(),
        kittyKeyboardFlags = live.kittyKeyboardFlags,
        // scrollback合成画面は毎回まったく別のセル内容(過去の行)を差し込むため、行単位の
        // dirty diff は意味を持たない。全画面dirty(null)にして初回/寸法変更時と同じく
        // グリッド全体を描き直させる(タスク#102)。
        dirtyRows = null,
    )
}
